### ISSUE TITLE
Module rights definition enhancement

### ISSUE THREAD

**Issue by hregis** [Dolibarr/dolibarr#29081]
### Feature Request

instead :

```
$this->rights[$r][0] = 1004;
$this->rights[$r][1] = 'Lire mouvements de stocks';
$this->rights[$r][2] = 'r';  // deprecated, à virer
$this->rights[$r][3] = 0;
$this->rights[$r][4] = 'mouvement';
$this->rights[$r][5] = 'lire'; 
```
it's better to use :

```
$this->rights[$r]['id'] = 1004;
$this->rights[$r]['label'] = 'Lire mouvements de stocks';
$this->rights[$r]['default'] = 0;
$this->rights[$r]['first_level'] = 'mouvement';
$this->rights[$r]['second_level'] = 'lire';
```
for to have possibility to add :

```
$this->rights[$r]['entity'] = 0;  // pour permettre d'ajouter des droits pour toutes les entités
$this->rights[$r]['element'] = 'invoice';  // pour permettre à un module externe d'ajouter des droits pour un autre module
$this->rights[$r]['visible'] = '$conf->global->MAIN_FEATURES_LEVEL >= 2';  // permettre une visibilité plus réactive
```
or

```
$this->rights[$r]['enable'] = '$conf->global->MAIN_FEATURES_LEVEL >= 2';  // permettre une visibilité plus réactive
```

or 

```
$this->rights[$r]['active'] = '$conf->global->MAIN_FEATURES_LEVEL >= 2';  // permettre une visibilité plus réactive
```

au choix ! ;-)

etc...

**Comment by JonBendtsen** [comment]
which file and version is your code from?

**Comment by hregis** [comment]
@JonBendtsen this is just an idea for the moment :-)

**Comment by mdeweerd** [comment]
As a complement to this suggestion, I propose to use more constants that are strings in stead of strings themselves.

```
$this->rights[$r]['id']
```

 would be

```
$this->rights[$r][Rights::KEY_ID] 
```

With Rights being the proper class where the constants are defined.

Hence, on the older versions these constants can be 0 to 5 and on the newer ones they can be the string values.
Also, this improves the detection of typing errors as constants need to exist where as strings can have any content.

Finally, 'label' should be in English.  I do not know have its translation is currently managed.  I can extend the script checking for translations by looking up the strings assigned directly to $this->rights[$r]['label'|1] vs. the translation - if I know where these translations are.

**Comment by JonBendtsen** [comment]
> @JonBendtsen this is just an idea for the moment :-)

none the less I like it

**Comment by hregis** [comment]
Work in progress : https://github.com/Dolibarr/dolibarr/pull/29150

**Comment by hregis** [comment]
> As a complement to this suggestion, I propose to use more constants that are strings in stead of strings themselves.
> 
> ```
> $this->rights[$r]['id']
> ```
> 
> would be
> 
> ```
> $this->rights[$r][Rights::KEY_ID] 
> ```
> 
> With Rights being the proper class where the constants are defined.
> 
> Hence, on the older versions these constants can be 0 to 5 and on the newer ones they can be the string values. Also, this improves the detection of typing errors as constants need to exist where as strings can have any content.
> 
> Finally, 'label' should be in English. I do not know have its translation is currently managed. I can extend the script checking for translations by looking up the strings assigned directly to $this->rights[$r]['label'|1] vs. the translation - if I know where these translations are.

Hi @mdeweerd 'label' is just an array key, no need to have a translation ? it's just so that it's more meaningful than an integer

**Comment by hregis** [comment]
@mdeweerd ok i see for this $this->rights[$r][Rights::KEY_ID]
good idea !

**Comment by hregis** [comment]
@mdeweerd 'label' in database is just for information (english or other, if translation exists, the translation is used... eg Permission5011, 5011 is the id of user right), the translation is after, in card
But it's possible to add the possibility to use translation key to have the default translation in database in the current language when the module is activated

**Comment by mdeweerd** [comment]
> @mdeweerd ok i see for this $this->rights[$r][Rights::KEY_ID] good idea !

Yes, that is more what I meant.  I thought that the value of $this->rights[$r]['label'] is the text shown to the user.  Either way, the value should be in English.

Rights::KEY_ID could be a number in the older dolibarr versions but a string in the new versions.

**Comment by mdeweerd** [comment]
For the record, I was also thinking of introducing `Right::READ => 'lire'` and similar so that the code would be more 'english' for the developer while maintaining the original key string.

**Comment by mdeweerd** [comment]
Regarding the test, this was inex 1 of rights:
> $this->rights[$r]['label'] = 'Lire mouvements de stocks';

So in the code that should be 'View or read the stock movements' .

**Comment by hregis** [comment]
@mdeweerd yes, little by little the bird is making its nest... I'm going to take it slowly so as not to break everything! ;-)

**Comment by hregis** [comment]
@mdeweerd for this "Rights::KEY_ID" the class "Rights" not exists

**Comment by hregis** [comment]
@mdeweerd @eldy maybe added a class "Rights" in user.class.php ?

**Comment by mdeweerd** [comment]
> "Rights::KEY_ID" the class "Rights" not exists

Yes 'Rights' was just an example.

It could be a separate class or User::RIGHTS_KEY_ID, User::RIGHTS_KEY_LABEL, etc  as well.  However, a dedicated class seems better.

**Comment by hregis** [comment]
@mdeweerd We're going to ask the big boss! @eldy will adding a "Rights" class make sense for later and use for the moment "Rights::KEY_ID" or can we just add "User::RIGHTS_KEY_ID" constants in the "User" class?

**Comment by hregis** [comment]
@mdeweerd look for "Rights::xxxx" https://github.com/Dolibarr/dolibarr/pull/29150/files

**Comment by mdeweerd** [comment]
> look for "Rights::xxxx" https://github.com/Dolibarr/dolibarr/pull/29150/files
Great!  I had a look and provided some feedback.

**Comment by eldy** [comment]
The idea of using Right::KEY_XXX is to increase code readability in module descriptors.
But it deems it will only be used in this file.
Other need is to read the table and in read mode, the key of table are used.
So why not using the constant into table DolibarrModule::KEY_XXX ?
